x/gamm/keeper/multihop_test.go

```diff
@@ -68,33 +102,48 @@ func (suite *KeeperTestSuite) TestBalancerPoolSimpleMultihopSwapExactAmountIn()
 						dec = dec.Mul(sp)
 					}
 					return dec
-				}()
+				}
 
-				tokenOutAmount, err := keeper.MultihopSwapExactAmountIn(suite.Ctx, suite.TestAccs[0], test.param.routes, test.param.tokenIn, test.param.tokenOutMinAmount)
-				suite.NoError(err, "test: %v", test.name)
+				// we create exact the same except swap fee 2 pool pairs
+				// use +2 to calc using second pair (w/o swap fee)
+				calcOutAmountAsSeparateSwaps := func(poolIdShift uint64) sdk.Coin {
+					cacheCtx, _ := suite.Ctx.CacheContext()
+					nextTokenIn := test.param.tokenIn
+					for _, hop := range test.param.routes {
 
-				// Calculate the chained spot price.
-				spotPriceAfter := func() sdk.Dec {
-					dec := sdk.NewDec(1)
-					tokenInDenom := test.param.tokenIn.Denom
-					for i, route := range test.param.routes {
-						if i != 0 {
-							tokenInDenom = test.param.routes[i-1].TokenOutDenom
-						}
+						tokenOut, err := keeper.SwapExactAmountIn(cacheCtx, suite.TestAccs[0], hop.PoolId+poolIdShift, nextTokenIn, hop.TokenOutDenom, sdk.OneInt())
+						suite.Require().NoError(err)
+						nextTokenIn = sdk.NewCoin(hop.TokenOutDenom, tokenOut)
+					}
+					return nextTokenIn
+				}
 
-						pool, err := keeper.GetPoolAndPoke(suite.Ctx, route.PoolId)
-						suite.NoError(err, "test: %v", test.name)
+				tokenOutCalculatedAsSeparateSwaps := calcOutAmountAsSeparateSwaps(0)
+				tokenOutCalculatedAsSeparateSwapsWithoutFee := calcOutAmountAsSeparateSwaps(2)
```

Comment:
I think what you are doing here is choosing 0 as the poolIdShift so you utilize the actual poolId which then utilizes the real fee. You then use 2 as the poolIdShift to select a non existent pool which has no fee because it doesn't exist.

What do you think about instead exporting the internal swapExactAmountIn (this https://github.com/osmosis-labs/osmosis/blob/c1808cd052160f3d187693b042cfd09892fc4f11/x/gamm/keeper/swap.go#L39-L40) via the export_test.go file (you would need to export it as some other name since the exported SwapExactAmountIn already exists). Idk, I am just thinking of ways to make this more robust / less hacky. Lmk what you think!

If you disagree with this method, I think at the very least we need to explain the hacky method we are doing in the comments (I know you said "use +2 to calc using second pair (w/o swap fee)" but without context this is confusing and took me a bit of time to figure out what was actually going on here)

Reply:
remade it with new logic of changing pool swap fee in cached context
previously it uses two pool pairs(what is not oblivious, i agree here.